packages/wrangler/src/pages/dev.ts

```diff
@@ -417,37 +429,126 @@ export const Handler = async (args: PagesDevArguments) => {
+					.filter(
+						(resolvedDep) =>
+							!resolvedDep.includes(normalize(singleWorkerScriptPath)) &&
+							!resolvedDep.includes(normalize("/.wrangler/")) &&
+							resolvedDep.match(resolve(process.cwd()))
+					);
+
+				// handle non-JS module dependencies, such as wasm/html/binary imports
+				const modules = bundle.modules
+					.filter((module) => !!module.filePath)
+					.map((module) =>
+						resolve(workerScriptDirectory, module.filePath as string)
+					);
+
+				/*
+				 *`bundle.dependencies` and `bundle.modules` combined, will always
+				 * provide us with the most up-to-date list of dependencies we need
+				 * to watch, since they reflect the latest built Worker bundle.
+				 * Therefore, we can safely unwatch all dependencies we have been
+				 * watching so far, and add all the new ones.
+				 */
+				currentBundleDependencies = [...dependencies, ...modules];
+
+				if (watchedBundleDependencies.length) {
+					watcher.unwatch(watchedBundleDependencies);
 				}
+				watcher.add(currentBundleDependencies);
+				watchedBundleDependencies = [...currentBundleDependencies];
 			};
 		}
 
-		await runBuild();
-		watch([workerScriptPath], {
-			persistent: true,
-			ignoreInitial: true,
-		}).on("all", async (event) => {
-			if (event === "unlink") {
-				return;
-			}
+		/*
+		 * Improve developer experience by debouncing the re-building
+		 * of the Worker script.
+		 *
+		 * "Debouncing ensures that exactly one signal is sent for an
+		 * event that may be happening several times — or even several
+		 * hundreds of times over an extended period. As long as the
+		 * events are occurring fast enough to happen at least once in
+		 * every detection period, the signal will not be sent!"
+		 * (http://unscriptable.com/2009/03/20/debouncing-javascript-methods/)
+		 *
+		 * This is great for the typing experience for example, when ideally
+		 * we want to re-build the Worker script once, when the user is done
+		 * typing, as opposed to once per each typed character.
+		 */
+		const debouncedRunBuild = (delayMS = 50) => {
+			const debounceFn = debounce(async () => {
+				try {
+					await runBuild();
+				} catch (e) {
+					/*
+					 * don't break developer flow in watch mode by throwing an error
+					 * here. Many times errors will be just the result of unfinished
+					 * typing. Instead, log the error, point out we are still serving
+					 * the last successfully built Worker, and allow developers to
+					 * write their code to completion
+					 */
+					logger.error(
+						`Error while attempting to build ${singleWorkerScriptPath}. Skipping to last successfully built version of the Worker.\n` +
+							`${e}`
+					);
+				}
+			}, delayMS);
+
+			debounceFn();
+		};
+
+		try {
 			await runBuild();
-		});
+
+			watcher.on("all", async (eventName, path) => {
+				logger.debug(`🌀 "${eventName}" event detected at ${path}.`);
+
+				// if "_worker.js" was deleted wait for 5s before re-building. It
```

Comment:
are we OK with this approach? and if yes, what's a realistic value for `delayMS`. I have abs no feel for it

Reply:
This feels a bit off to me, and I think framework users would probably be surprised that that their rebuilt changes take 5 seconds to show up. What about instead watching the parent directory of `_worker.js`, so we can immediately react to the creation event for `_worker.js`?

Reply:
Why rebuild at all, if they have deleted? Seems like this behaviour would be unexpected.

Reply:
then perhaps the better question here is do we want to rebuild if `_worker.js` has been deleted? For me it was unexpected that nothing happened when I deleted the file. But maybe that's just me?

Reply:
> users would probably be surprised that that their rebuilt changes take 5 seconds to show up
to clarify this will only happen when the `_worker.js` file is deleted, **not** on every change

> What about instead watching the parent directory of _worker.js
I might be missing what you have in mind, but how would this solve the problem at hand? The issue is not reacting to the adding of the file, which this implementation already does, it's the adding it "immediately" (not sure what that actually looks like in terms of `ms`) after it was deleted, and either debouncing the `unlink` event handler, or bypassing it completely.

Reply:
Maybe I'd misunderstand, but I thought the problem this was solving was that the `_worker.js` is deleted by some frameworks on every reload?

Reply:
that is correct yes. and because said frameworks also add the file back, we'd ideally want to either:
- delay building the worker until after the file is added back (which is what this PR does)
- or by-pass rebuilding the worker completely whenever worker file gets deleted (which is how it currently works)

your proposal to react immediately to the `addFile` event (by watching the parent folder), is something that is already accomplished in the crr PR by watching the file itself. The problem in both cases would remain that we'd get all sorts of compile errors when attempting to build the worker after the file was deleted, unless we do one of the 👆

Reply:
What I'm concerned about is:
- Framework deletes `_worker.js` file
- We delay building for 5 seconds
- Framework re-adds `_worker.js` file as part of the build process
- User is waiting for 5 seconds for the build to trigger

Would that not happen?

Reply:
yes that can happen if the framework adds the file back before those 5s have elapsed. So say file is deleted, then added after 1s...the user needs to wait for 4 more seconds before we trigger a rebuild. But this is why I was saying that I have zero feels abt how long this delay should be. We can set it to whatever we want.

Reply:
At the same time, if the framework re-adds the file after 7s, we trigger a rebuild after 5s, which will correctly error because the file does not exist, then trigger another build after 2s when the file was re-added, which, assuming all's well, will compile successfully. So in this case, even with a debounce in place, we'd still run into rebuilding a worker file that does not exist, which is what users don't want.

Reply:
So perhaps, instead of trying to get that delay right, which I'm guessing is impossible, cause the frameworks build process duration is relative to the project size, maybe keep the current behaviour and nor rebuild whenever the file is deleted? I still find the lack of feedback odd, but it might possibly be just me

Reply:
> maybe keep the current behaviour and not rebuild whenever the file is deleted?
I think this might be best, even if it is a bit confusing, since frameworks seem to rely on this behaviour. Maybe we could log a small warning to the terminal?